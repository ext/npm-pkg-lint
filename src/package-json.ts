import { type PackageJson } from "./types";
import { type Message } from "./message";
import { type Result } from "./result";
import { nonempty, present, typeArray, typeString, ValidationError, validUrl } from "./validators";
import { isDisallowedDependency } from "./rules/disallowed-dependency";
import { exportsTypesOrder } from "./rules/exports-types-order";
import { outdatedEngines } from "./rules/outdated-engines";
import { verifyEngineConstraint } from "./rules/verify-engine-constraint";
import { typesNodeMatchingEngine } from "./rules/types-node-matching-engine";

export interface VerifyPackageJsonOptions {
	allowedDependencies: Set<string>;
	allowTypesDependencies?: boolean;
	ignoreMissingFields?: boolean;
}

type validator = (key: string, value: unknown) => void;

const fields: Record<string, validator[]> = {
	description: [present, typeString, nonempty],
	keywords: [present, typeArray, nonempty],
	homepage: [present, typeString, validUrl],
	bugs: [present, validUrl],
	license: [present, typeString, nonempty],
	author: [present, nonempty],
	repository: [present, validUrl],
};

function verifyFields(pkg: PackageJson, options: VerifyPackageJsonOptions): Message[] {
	const messages: Message[] = [];

	for (const [field, validators] of Object.entries(fields)) {
		try {
			for (const validator of validators) {
				validator(field, pkg[field]);
			}
		} catch (error) {
			if (
				error instanceof ValidationError &&
				error.validator === present.name &&
				options.ignoreMissingFields
			) {
				continue;
			}
			// istanbul ignore else
			if (error instanceof Error) {
				messages.push({
					ruleId: "package-json-fields",
					severity: 2,
					message: error.message,
					line: 1,
					column: 1,
				});
			} else {
				throw error;
			}
		}
	}

	return messages;
}

function getActualDependency(key: string, version: string): string {
	/* handle npm: prefix */
	if (version.startsWith("npm:")) {
		const [name] = version.slice("npm:".length).split("@", 2);
		return name;
	}

	return key;
}

function verifyDependencies(pkg: PackageJson, options: VerifyPackageJsonOptions): Message[] {
	const messages: Message[] = [];

	for (const [key, version] of Object.entries(pkg.dependencies ?? {})) {
		const dependency = getActualDependency(key, version);

		/* skip dependencies explicitly allowed by the user */
		if (options.allowedDependencies.has(dependency)) {
			continue;
		}

		/* skip @types/* if explicitly allowed by user */
		if (options.allowTypesDependencies && dependency.match(/^@types\//)) {
			continue;
		}

		if (isDisallowedDependency(pkg, dependency)) {
			const name = key === dependency ? dependency : `"${key}" ("npm:${dependency}")`;
			messages.push({
				ruleId: "disallowed-dependency",
				severity: 2,
				message: `"${name}" should be a devDependency`,
				line: 1,
				column: 1,
			});
		}
	}

	return messages;
}

export async function verifyPackageJson(
	pkg: PackageJson,
	filePath: string,
	options: VerifyPackageJsonOptions = { allowedDependencies: new Set() }
): Promise<Result[]> {
	const messages: Message[] = [
		...(await verifyEngineConstraint(pkg)),
		...exportsTypesOrder(pkg),
		...verifyFields(pkg, options),
		...verifyDependencies(pkg, options),
		...outdatedEngines(pkg),
		...typesNodeMatchingEngine(pkg),
	];

	if (messages.length === 0) {
		return [];
	}

	return [
		{
			messages,
			filePath,
			errorCount: messages.filter((it) => it.severity === 2).length,
			warningCount: messages.filter((it) => it.severity === 1).length,
			fixableErrorCount: 0,
			fixableWarningCount: 0,
		},
	];
}
