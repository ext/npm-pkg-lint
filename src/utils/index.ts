export { type NpmInfoError, isNpmInfoError, npmInfo } from "./npm-info";
