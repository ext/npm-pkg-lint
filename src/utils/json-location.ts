import { type DocumentNode, type MemberNode, type ValueNode } from "@humanwhocodes/momoa";

interface Result {
	value: ValueNode;
	member?: MemberNode;
}

function findNextNode(current: Result, segment: string, datapath: string[]): Result {
	const { value } = current;
	switch (value.type) {
		case "Object": {
			const found = value.members.find((it) => {
				return it.name.type === "String" && it.name.value === segment;
			});
			if (!found) {
				throw new Error(`Failed to find "${datapath.join(".")}" in JSON structure`);
			}
			return { value: found.value, member: found };
		}
		default: {
			throw new Error(
				`Dont know how to handle node type "${value.type}" when searching for "${datapath.join(
					".",
				)}" in JSON structure`,
			);
		}
	}
}

export function jsonLocation(
	ast: DocumentNode,
	pick: "member" | "value",
	...datapath: string[]
): { line: number; column: number } {
	const result = datapath.reduce<Result>(
		(state, segment) => findNextNode(state, segment, datapath),
		{ value: ast.body },
	);

	switch (pick) {
		case "member":
			return result.member?.loc.start ?? { line: 1, column: 1 };
		case "value":
			return result.value.loc.start;
	}
}
