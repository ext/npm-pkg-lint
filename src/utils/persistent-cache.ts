import fs from "node:fs/promises";
import path from "node:path";
import crypto from "node:crypto";
import os from "node:os";

const enabled = os.platform() === "linux";
const cacheRoot = process.env.XDG_CACHE_HOME ?? path.join(os.homedir(), ".cache");
let cacheDir = path.join(cacheRoot, "npm-pkg-lint");

/* istanbul ignore next */
function getFilePath(key: string): string {
	const hash = crypto.createHash("md5").update(key).digest("hex");
	const filename = `${hash.slice(0, 2)}/${hash.slice(2)}.json`;
	return path.join(cacheDir, filename);
}

/* istanbul ignore next */
export async function setCacheDirecory(directory: string): Promise<void> {
	await fs.mkdir(directory, { recursive: true });
	cacheDir = directory;
}

/* istanbul ignore next */
export async function persistentCacheGet(key: string): Promise<unknown | null> {
	if (!enabled) {
		return null;
	}

	const filePath = getFilePath(key);
	try {
		const content = await fs.readFile(filePath, "utf-8");
		return JSON.parse(content) as unknown;
	} catch (err: unknown) {
		if (err instanceof Error && "code" in err && err.code === "ENOENT") {
			return null;
		}
		throw err;
	}
}

/* istanbul ignore next */
export async function persistentCacheSet(key: string, data: unknown): Promise<void> {
	if (!enabled) {
		return;
	}

	const filePath = getFilePath(key);
	const content = JSON.stringify(data, null, 2);
	await fs.mkdir(path.dirname(filePath), { recursive: true });
	await fs.writeFile(filePath, content, "utf-8");
}
