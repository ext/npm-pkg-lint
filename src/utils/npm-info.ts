import { execa } from "execa";
import { type PackageJson } from "../types";
import { persistentCacheGet, persistentCacheSet } from "./persistent-cache";

export interface NpmInfoError {
	code: string;
	summary: string;
	detail: string;
}

export interface ExecaError extends Error {
	stdout: string;
}

const cache = new Map<string, PackageJson | null>();

function isExecaError(error: unknown): error is ExecaError {
	return Boolean(error && error instanceof Error && "stdout" in error);
}

export function isNpmInfoError(error: unknown): error is NpmInfoError {
	return Boolean(error && error instanceof Error && "summary" in error);
}

function tryParse(maybeJson: string): { error: NpmInfoError } | null {
	try {
		return JSON.parse(maybeJson) as { error: NpmInfoError };
	} catch {
		return null;
	}
}

export async function npmInfo(pkg: string): Promise<PackageJson>;
export async function npmInfo(
	pkg: string,
	options: { ignoreUnpublished: true }
): Promise<PackageJson | null>;
export async function npmInfo(
	pkg: string,
	options: { ignoreUnpublished: boolean } = { ignoreUnpublished: false }
): Promise<PackageJson | null> {
	const { ignoreUnpublished } = options;
	const cached = cache.get(pkg);
	if (cached === null) {
		if (ignoreUnpublished) {
			return null;
		}
	} else if (cached) {
		return cached;
	}

	const persistent = await persistentCacheGet(pkg);
	if (persistent) {
		return persistent as PackageJson;
	}

	try {
		const result = await execa("npm", ["info", "--json", pkg]);
		const pkgData = JSON.parse(result.stdout) as PackageJson;
		cache.set(pkg, pkgData);
		await persistentCacheSet(pkg, pkgData);
		return pkgData;
	} catch (err: unknown) {
		if (!isExecaError(err)) {
			throw err;
		}
		const parsed = tryParse(err.stdout);
		if (!parsed) {
			throw err;
		}
		const { code, summary, detail } = parsed.error;

		/* cache for this session but don't store in persistent cache as this
		 * error might be temporary */
		cache.set(pkg, null);

		if (ignoreUnpublished && code === "E404") {
			return null;
		}

		const wrappedError = new Error(summary) as Error & NpmInfoError;
		wrappedError.code = code;
		wrappedError.summary = summary;
		wrappedError.detail = detail;
		throw wrappedError;
	}
}
