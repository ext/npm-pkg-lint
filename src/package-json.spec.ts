import { parse, type DocumentNode } from "@humanwhocodes/momoa";
import { verifyPackageJson } from "./package-json";
import { PackageJson } from "./types";
import { codeframe } from "./utils/codeframe";

// eslint-disable-next-line @typescript-eslint/ban-ts-comment -- it exists in the mock only
// @ts-expect-error
import { npmInfoMockDefault } from "./utils/npm-info";

jest.mock("./utils/npm-info");

let pkg: PackageJson & Required<Pick<PackageJson, "engines">>;

function generateAst(pkg: PackageJson): { content: string; ast: DocumentNode } {
	const content = JSON.stringify(pkg, null, 2);
	return { content, ast: parse(content) };
}

beforeEach(() => {
	pkg = {
		name: "mock-pkg",
		version: "1.2.3",
		description: "description",
		keywords: ["test"],
		homepage: "https://example.net",
		bugs: "https://example.net",
		license: "UNLICENSED",
		author: "Fred Flintstone <fred.flintstone@example.net>",
		repository: { type: "git", url: "git+https://git.example.net/test-case.git" },
		engines: {
			node: ">= 20",
		},
	};
	npmInfoMockDefault(pkg);
});

it("should not return errors if package.json is well formed (strings only)", async () => {
	expect.assertions(1);
	const { ast } = generateAst(pkg);
	const results = await verifyPackageJson(pkg, ast, "package.json");
	expect(results).toEqual([]);
});

it("should not return errors if package.json has no dependencies", async () => {
	expect.assertions(1);
	delete pkg.dependencies;
	const { ast } = generateAst(pkg);
	const results = await verifyPackageJson(pkg, ast, "package.json");
	expect(results).toEqual([]);
});

it("should return error if dependency is disallowed", async () => {
	expect.assertions(1);
	pkg.dependencies = {
		eslint: "1.2.3",
	};
	const { content, ast } = generateAst(pkg);
	const results = await verifyPackageJson(pkg, ast, "package.json");
	expect(codeframe(content, results)).toMatchInlineSnapshot(`
		"ERROR: "eslint" should be a devDependency (disallowed-dependency) at package.json
		  18 |   },
		  19 |   "dependencies": {
		> 20 |     "eslint": "1.2.3"
		     |     ^
		  21 |   }
		  22 | }"
	`);
});

it("should return error if aliased dependency is disallowed", async () => {
	expect.assertions(1);
	pkg.dependencies = {
		aliased: "npm:eslint@1.2.3",
	};
	const { content, ast } = generateAst(pkg);
	const results = await verifyPackageJson(pkg, ast, "package.json");
	expect(codeframe(content, results)).toMatchInlineSnapshot(`
		"ERROR: "aliased" ("npm:eslint") should be a devDependency (disallowed-dependency) at package.json
		  18 |   },
		  19 |   "dependencies": {
		> 20 |     "aliased": "npm:eslint@1.2.3"
		     |     ^
		  21 |   }
		  22 | }"
	`);
});

it("should return error if dependency is obsolete", async () => {
	expect.assertions(1);
	pkg.devDependencies = {
		mkdirp: "1.2.3",
	};
	const { content, ast } = generateAst(pkg);
	const results = await verifyPackageJson(pkg, ast, "package.json");
	expect(codeframe(content, results)).toMatchInlineSnapshot(`
		"ERROR: "mkdirp" is obsolete and should no longer be used: use native "fs.mkdir(..., { recursive: true })" instead (obsolete-dependency) at package.json
		  18 |   },
		  19 |   "devDependencies": {
		> 20 |     "mkdirp": "1.2.3"
		     |     ^
		  21 |   }
		  22 | }"
	`);
});

it("should not return error if dependency is allowed", async () => {
	expect.assertions(1);
	pkg.dependencies = {
		allowed: "1.2.3",
	};
	const { ast } = generateAst(pkg);
	const results = await verifyPackageJson(pkg, ast, "package.json");
	expect(results).toEqual([]);
});

it("should not return error if explicitly allowed by user", async () => {
	expect.assertions(1);
	pkg.dependencies = {
		eslint: "1.2.3",
		aliased: "npm:eslint@1.2.3",
	};
	const { ast } = generateAst(pkg);
	const results = await verifyPackageJson(pkg, ast, "package.json", {
		allowedDependencies: new Set(["eslint"]),
		ignoreNodeVersion: false,
	});
	expect(results).toHaveLength(0);
});

it("should return engines.node supports eol version", async () => {
	expect.assertions(1);
	pkg.engines.node = ">= 8";
	const { content, ast } = generateAst(pkg);
	const results = await verifyPackageJson(pkg, ast, "package.json");
	expect(codeframe(content, results)).toMatchInlineSnapshot(`
		"ERROR: engines.node is satisfied by Node 8 (EOL since 2019-12-31) (outdated-engines) at package.json
		  15 |   },
		  16 |   "engines": {
		> 17 |     "node": ">= 8"
		     |             ^
		  18 |   }
		  19 | }"
	`);
});

describe("@types", () => {
	beforeEach(() => {
		pkg.dependencies = {
			"@types/foobar": "1.2.3",
		};
	});

	it("should return error if @types is used as dependency by default", async () => {
		expect.assertions(1);
		const { content, ast } = generateAst(pkg);
		const results = await verifyPackageJson(pkg, ast, "package.json");
		expect(codeframe(content, results)).toMatchInlineSnapshot(`
			"ERROR: "@types/foobar" should be a devDependency (disallowed-dependency) at package.json
			  18 |   },
			  19 |   "dependencies": {
			> 20 |     "@types/foobar": "1.2.3"
			     |     ^
			  21 |   }
			  22 | }"
		`);
	});

	it("should not return error if @types is allowed", async () => {
		expect.assertions(1);
		const { ast } = generateAst(pkg);
		const results = await verifyPackageJson(pkg, ast, "package.json", {
			allowedDependencies: new Set(),
			allowTypesDependencies: true,
			ignoreNodeVersion: false,
		});
		expect(results).toHaveLength(0);
	});
});

describe("present", () => {
	it("should return error on missing fields by default", async () => {
		expect.assertions(1);
		delete pkg.description;
		const { content, ast } = generateAst(pkg);
		const results = await verifyPackageJson(pkg, ast, "package.json");
		expect(codeframe(content, results)).toMatchInlineSnapshot(`
			"ERROR: "description" must be set (package-json-fields) at package.json
			> 1 | {
			    | ^
			  2 |   "name": "mock-pkg",
			  3 |   "version": "1.2.3",
			  4 |   "keywords": ["
		`);
	});

	it("should ignore error on missing fields if ignoreMissingFields is set", async () => {
		expect.assertions(1);
		delete pkg.description;
		const { ast } = generateAst(pkg);
		const results = await verifyPackageJson(pkg, ast, "package.json", {
			allowedDependencies: new Set(),
			ignoreMissingFields: true,
			ignoreNodeVersion: false,
		});
		expect(results).toHaveLength(0);
	});
});

describe("fields", () => {
	describe("description", () => {
		it("should return error if not set", async () => {
			expect.assertions(1);
			delete pkg.description;
			const { content, ast } = generateAst(pkg);
			const results = await verifyPackageJson(pkg, ast, "package.json");
			expect(codeframe(content, results)).toMatchInlineSnapshot(`
				"ERROR: "description" must be set (package-json-fields) at package.json
				> 1 | {
				    | ^
				  2 |   "name": "mock-pkg",
				  3 |   "version": "1.2.3",
				  4 |   "keywords": ["
			`);
		});

		it("should return error if not string", async () => {
			expect.assertions(1);
			pkg.description = 12 as unknown as string;
			const { content, ast } = generateAst(pkg);
			const results = await verifyPackageJson(pkg, ast, "package.json");
			expect(codeframe(content, results)).toMatchInlineSnapshot(`
				"ERROR: "description" must be string (package-json-fields) at package.json
				  2 |   "name": "mock-pkg",
				  3 |   "version": "1.2.3",
				> 4 |   "description": 12,
				    |                  ^
				  5 |   "keywords": [
				  6 |     "test"
				  7 |   ],"
			`);
		});

		it("should return error if empty", async () => {
			expect.assertions(1);
			pkg.description = "";
			const { content, ast } = generateAst(pkg);
			const results = await verifyPackageJson(pkg, ast, "package.json");
			expect(codeframe(content, results)).toMatchInlineSnapshot(`
				"ERROR: "description" must not be empty (package-json-fields) at package.json
				  2 |   "name": "mock-pkg",
				  3 |   "version": "1.2.3",
				> 4 |   "description": "",
				    |                  ^
				  5 |   "keywords": [
				  6 |     "test"
				  7 |   ],"
			`);
		});
	});

	describe("keywords", () => {
		it("should return error if not set", async () => {
			expect.assertions(1);
			delete pkg.keywords;
			const { content, ast } = generateAst(pkg);
			const results = await verifyPackageJson(pkg, ast, "package.json");
			expect(codeframe(content, results)).toMatchInlineSnapshot(`
				"ERROR: "keywords" must be set (package-json-fields) at package.json
				> 1 | {
				    | ^
				  2 |   "name": "mock-pkg",
				  3 |   "version": "1.2.3",
				  4 |   "description": "description","
			`);
		});

		it("should return error if not string", async () => {
			expect.assertions(1);
			pkg.keywords = 12 as unknown as string[];
			const { content, ast } = generateAst(pkg);
			const results = await verifyPackageJson(pkg, ast, "package.json");
			expect(codeframe(content, results)).toMatchInlineSnapshot(`
				"ERROR: "keywords" must be array (package-json-fields) at package.json
				  3 |   "version": "1.2.3",
				  4 |   "description": "description",
				> 5 |   "keywords": 12,
				    |               ^
				  6 |   "homepage": "https://example.net",
				  7 |   "bugs": "https://example.net",
				  8 |   "license": "UNLICENSED","
			`);
		});

		it("should return error if empty", async () => {
			expect.assertions(1);
			pkg.keywords = [];
			const { content, ast } = generateAst(pkg);
			const results = await verifyPackageJson(pkg, ast, "package.json");
			expect(codeframe(content, results)).toMatchInlineSnapshot(`
				"ERROR: "keywords" must not be empty (package-json-fields) at package.json
				  3 |   "version": "1.2.3",
				  4 |   "description": "description",
				> 5 |   "keywords": [],
				    |               ^
				  6 |   "homepage": "https://example.net",
				  7 |   "bugs": "https://example.net",
				  8 |   "license": "UNLICENSED","
			`);
		});

		it("should return error if items are empty", async () => {
			expect.assertions(1);
			pkg.keywords = [""];
			const { content, ast } = generateAst(pkg);
			const results = await verifyPackageJson(pkg, ast, "package.json");
			expect(codeframe(content, results)).toMatchInlineSnapshot(`
				"ERROR: "keywords[0]" must not be empty (package-json-fields) at package.json
				  3 |   "version": "1.2.3",
				  4 |   "description": "description",
				> 5 |   "keywords": [
				    |               ^
				  6 |     ""
				  7 |   ],
				  8 |   "homepage": "https://example.net","
			`);
		});
	});

	describe("homepage", () => {
		it("should return error if not set", async () => {
			expect.assertions(1);
			delete pkg.homepage;
			const { content, ast } = generateAst(pkg);
			const results = await verifyPackageJson(pkg, ast, "package.json");
			expect(codeframe(content, results)).toMatchInlineSnapshot(`
				"ERROR: "homepage" must be set (package-json-fields) at package.json
				> 1 | {
				    | ^
				  2 |   "name": "mock-pkg",
				  3 |   "version": "1.2.3",
				  4 |   "description": "description","
			`);
		});

		it("should return error if not string", async () => {
			expect.assertions(1);
			pkg.homepage = 12 as unknown as string;
			const { content, ast } = generateAst(pkg);
			const results = await verifyPackageJson(pkg, ast, "package.json");
			expect(codeframe(content, results)).toMatchInlineSnapshot(`
				"ERROR: "homepage" must be string (package-json-fields) at package.json
				   6 |     "test"
				   7 |   ],
				>  8 |   "homepage": 12,
				     |               ^
				   9 |   "bugs": "https://example.net",
				  10 |   "license": "UNLICENSED",
				  11 |   "author": "Fred Flintstone <fred.flintstone@example.net>","
			`);
		});

		it("should return error if not valid url", async () => {
			expect.assertions(1);
			pkg.homepage = "foobar";
			const { content, ast } = generateAst(pkg);
			const results = await verifyPackageJson(pkg, ast, "package.json");
			expect(codeframe(content, results)).toMatchInlineSnapshot(`
				"ERROR: "homepage" must be a valid url (https only) (package-json-fields) at package.json
				   6 |     "test"
				   7 |   ],
				>  8 |   "homepage": "foobar",
				     |               ^
				   9 |   "bugs": "https://example.net",
				  10 |   "license": "UNLICENSED",
				  11 |   "author": "Fred Flintstone <fred.flintstone@example.net>","
			`);
		});
	});

	describe("bugs", () => {
		it("should return error if not set", async () => {
			expect.assertions(1);
			delete pkg.bugs;
			const { content, ast } = generateAst(pkg);
			const results = await verifyPackageJson(pkg, ast, "package.json");
			expect(codeframe(content, results)).toMatchInlineSnapshot(`
				"ERROR: "bugs" must be set (package-json-fields) at package.json
				> 1 | {
				    | ^
				  2 |   "name": "mock-pkg",
				  3 |   "version": "1.2.3",
				  4 |   "description": "description","
			`);
		});

		it("should return error if not valid url", async () => {
			expect.assertions(1);
			pkg.bugs = "foobar";
			const { content, ast } = generateAst(pkg);
			const results = await verifyPackageJson(pkg, ast, "package.json");
			expect(codeframe(content, results)).toMatchInlineSnapshot(`
				"ERROR: "bugs" must be a valid url (https only) (package-json-fields) at package.json
				   7 |   ],
				   8 |   "homepage": "https://example.net",
				>  9 |   "bugs": "foobar",
				     |           ^
				  10 |   "license": "UNLICENSED",
				  11 |   "author": "Fred Flintstone <fred.flintstone@example.net>",
				  12 |   "repository": {"
			`);
		});
	});

	describe("license", () => {
		it("should return error if not set", async () => {
			expect.assertions(1);
			delete pkg.license;
			const { content, ast } = generateAst(pkg);
			const results = await verifyPackageJson(pkg, ast, "package.json");
			expect(codeframe(content, results)).toMatchInlineSnapshot(`
				"ERROR: "license" must be set (package-json-fields) at package.json
				> 1 | {
				    | ^
				  2 |   "name": "mock-pkg",
				  3 |   "version": "1.2.3",
				  4 |   "description": "description","
			`);
		});

		it("should return error if not string", async () => {
			expect.assertions(1);
			pkg.license = 12 as unknown as string;
			const { content, ast } = generateAst(pkg);
			const results = await verifyPackageJson(pkg, ast, "package.json");
			expect(codeframe(content, results)).toMatchInlineSnapshot(`
				"ERROR: "license" must be string (package-json-fields) at package.json
				   8 |   "homepage": "https://example.net",
				   9 |   "bugs": "https://example.net",
				> 10 |   "license": 12,
				     |              ^
				  11 |   "author": "Fred Flintstone <fred.flintstone@example.net>",
				  12 |   "repository": {
				  13 |     "type": "git","
			`);
		});

		it("should return error if empty", async () => {
			expect.assertions(1);
			pkg.license = "";
			const { content, ast } = generateAst(pkg);
			const results = await verifyPackageJson(pkg, ast, "package.json");
			expect(codeframe(content, results)).toMatchInlineSnapshot(`
				"ERROR: "license" must not be empty (package-json-fields) at package.json
				   8 |   "homepage": "https://example.net",
				   9 |   "bugs": "https://example.net",
				> 10 |   "license": "",
				     |              ^
				  11 |   "author": "Fred Flintstone <fred.flintstone@example.net>",
				  12 |   "repository": {
				  13 |     "type": "git","
			`);
		});
	});

	describe("author", () => {
		it("should return error if not set", async () => {
			expect.assertions(1);
			delete pkg.author;
			const { content, ast } = generateAst(pkg);
			const results = await verifyPackageJson(pkg, ast, "package.json");
			expect(codeframe(content, results)).toMatchInlineSnapshot(`
				"ERROR: "author" must be set (package-json-fields) at package.json
				> 1 | {
				    | ^
				  2 |   "name": "mock-pkg",
				  3 |   "version": "1.2.3",
				  4 |   "description": "description","
			`);
		});

		it("should return error if empty string", async () => {
			expect.assertions(1);
			pkg.author = "";
			const { content, ast } = generateAst(pkg);
			const results = await verifyPackageJson(pkg, ast, "package.json");
			expect(codeframe(content, results)).toMatchInlineSnapshot(`
				"ERROR: "author" must not be empty (package-json-fields) at package.json
				   9 |   "bugs": "https://example.net",
				  10 |   "license": "UNLICENSED",
				> 11 |   "author": "",
				     |             ^
				  12 |   "repository": {
				  13 |     "type": "git",
				  14 |     "url": "git+https://git.example.net/test-case.git""
			`);
		});

		it("should return error if empty object", async () => {
			expect.assertions(1);
			pkg.author = {};
			const { content, ast } = generateAst(pkg);
			const results = await verifyPackageJson(pkg, ast, "package.json");
			expect(codeframe(content, results)).toMatchInlineSnapshot(`
				"ERROR: "author" must not be empty (package-json-fields) at package.json
				   9 |   "bugs": "https://example.net",
				  10 |   "license": "UNLICENSED",
				> 11 |   "author": {},
				     |             ^
				  12 |   "repository": {
				  13 |     "type": "git",
				  14 |     "url": "git+https://git.example.net/test-case.git""
			`);
		});
	});

	describe("repository", () => {
		it("should return error if not set", async () => {
			expect.assertions(1);
			delete pkg.repository;
			const { content, ast } = generateAst(pkg);
			const results = await verifyPackageJson(pkg, ast, "package.json");
			expect(codeframe(content, results)).toMatchInlineSnapshot(`
				"ERROR: "repository" must be set (package-json-fields) at package.json
				> 1 | {
				    | ^
				  2 |   "name": "mock-pkg",
				  3 |   "version": "1.2.3",
				  4 |   "description": "description","
			`);
		});

		it("should return error if not not an object", async () => {
			expect.assertions(1);
			pkg.repository = "foobar";
			const { content, ast } = generateAst(pkg);
			const results = await verifyPackageJson(pkg, ast, "package.json");
			expect(codeframe(content, results)).toMatchInlineSnapshot(`
				"ERROR: "repository" must be an object with "type" and "url" (package-json-fields) at package.json
				  10 |   "license": "UNLICENSED",
				  11 |   "author": "Fred Flintstone <fred.flintstone@example.net>",
				> 12 |   "repository": "foobar",
				     |                 ^
				  13 |   "engines": {
				  14 |     "node": ">= 20"
				  15 |   }"
			`);
		});

		it("should return error if not valid url", async () => {
			expect.assertions(1);
			pkg.repository = { type: "git", url: "http://example.net" };
			const { content, ast } = generateAst(pkg);
			const results = await verifyPackageJson(pkg, ast, "package.json");
			expect(codeframe(content, results)).toMatchInlineSnapshot(`
				"ERROR: "repository.url" must use "git+https://" instead of "http://" (package-json-fields) at package.json
				  10 |   "license": "UNLICENSED",
				  11 |   "author": "Fred Flintstone <fred.flintstone@example.net>",
				> 12 |   "repository": {
				     |                 ^
				  13 |     "type": "git",
				  14 |     "url": "http://example.net"
				  15 |   },"
			`);
		});
	});
});
