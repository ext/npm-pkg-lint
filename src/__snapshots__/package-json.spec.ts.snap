// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fields author should return error if empty object 1`] = `
[
  {
    "column": 1,
    "line": 1,
    "message": ""author" must not be empty",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`fields author should return error if empty string 1`] = `
[
  {
    "column": 1,
    "line": 1,
    "message": ""author" must not be empty",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`fields author should return error if not set 1`] = `
[
  {
    "column": 1,
    "line": 1,
    "message": ""author" must be set",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`fields bugs should return error if not set 1`] = `
[
  {
    "column": 1,
    "line": 1,
    "message": ""bugs" must be set",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`fields bugs should return error if not valid url 1`] = `
[
  {
    "column": 1,
    "line": 1,
    "message": ""bugs" must be a valid url (https only)",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`fields description should return error if empty 1`] = `
[
  {
    "column": 1,
    "line": 1,
    "message": ""description" must not be empty",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`fields description should return error if not set 1`] = `
[
  {
    "column": 1,
    "line": 1,
    "message": ""description" must be set",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`fields description should return error if not string 1`] = `
[
  {
    "column": 1,
    "line": 1,
    "message": ""description" must be string",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`fields homepage should return error if not set 1`] = `
[
  {
    "column": 1,
    "line": 1,
    "message": ""homepage" must be set",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`fields homepage should return error if not string 1`] = `
[
  {
    "column": 1,
    "line": 1,
    "message": ""homepage" must be string",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`fields homepage should return error if not valid url 1`] = `
[
  {
    "column": 1,
    "line": 1,
    "message": ""homepage" must be a valid url (https only)",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`fields keywords should return error if empty 1`] = `
[
  {
    "column": 1,
    "line": 1,
    "message": ""keywords" must not be empty",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`fields keywords should return error if items are empty 1`] = `
[
  {
    "column": 1,
    "line": 1,
    "message": ""keywords[0]" must not be empty",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`fields keywords should return error if not set 1`] = `
[
  {
    "column": 1,
    "line": 1,
    "message": ""keywords" must be set",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`fields keywords should return error if not string 1`] = `
[
  {
    "column": 1,
    "line": 1,
    "message": ""keywords" must be array",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`fields license should return error if empty 1`] = `
[
  {
    "column": 1,
    "line": 1,
    "message": ""license" must not be empty",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`fields license should return error if not set 1`] = `
[
  {
    "column": 1,
    "line": 1,
    "message": ""license" must be set",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`fields license should return error if not string 1`] = `
[
  {
    "column": 1,
    "line": 1,
    "message": ""license" must be string",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`fields repository should return error if not set 1`] = `
[
  {
    "column": 1,
    "line": 1,
    "message": ""repository" must be set",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`fields repository should return error if not valid url 1`] = `
[
  {
    "column": 1,
    "line": 1,
    "message": ""repository" must be a valid url (https only)",
    "ruleId": "package-json-fields",
    "severity": 2,
  },
]
`;

exports[`should return engines.node supports eol version 1`] = `
[
  {
    "column": 1,
    "line": 1,
    "message": "engines.node is satisfied by Node 8 (EOL since 2019-12-31)",
    "ruleId": "outdated-engines",
    "severity": 2,
  },
]
`;
